src/v/redpanda/application.h

```diff
@@ -315,6 +315,48 @@ class application {
         }
     }
 
+    // Shutdown the service with custom method.
+    // The method should be invoked in the ss::thread context.
+    template<class StopFunc>
+    void shutdown_with_watchdog(
+      ss::sstring name, ss::sstring method_name, StopFunc&& stop_func) {
+        // This watchdog is triggered after short period of time (30s). It
+        // adds message to the log that service is taking a long time to
+        // shutdown on INFO level.
+        ssx::watchdog short_wd(
+          short_shutdown_warning_timeout, [this, name, method_name] {
+              vlog(
+                _log.info,
+                "Service {} is taking more than {} seconds to shutdown with "
+                "{}.",
+                name,
+                std::chrono::duration_cast<std::chrono::seconds>(
+                  short_shutdown_warning_timeout)
+                  .count(),
+                method_name);
+          });
+        // This watchdog is triggered after long period of time. This indicates
+        // a bug (most likely).
+        ssx::watchdog long_wd(
+          long_shutdown_warning_timeout, [this, name, method_name] {
+              vlog(
+                _log.error,
+                "Service {} is taking more than {} seconds to shutdown with "
+                "{}!",
+                name,
+                std::chrono::duration_cast<std::chrono::seconds>(
+                  long_shutdown_warning_timeout)
+                  .count(),
+                method_name);
+          });
+
+        vlog(_log.info, "Shutting down {} with {} method", name, method_name);
+
+        ss::futurize_invoke(std::forward<StopFunc>(stop_func)).get();
```

Comment:
Why do we need `ss::futurize_invoke` here?
We're not after converting an immediate exception into an exceptional future here, as there's no chain or coroutine.

And if `stop_func` is synchronous watchdogging it does not make sense, so it's better to result in a compilation error.
The only exception is functions that are not synchronous but are not coroutines: those that use .get() inside -- I guess we avoid them other than on the very top level?